### PR TITLE
Simplify config file to be more user friendly

### PR DESCRIPTION
**Pull request by varunshenoy**
Resolving the issue of simplifying the config as shared in #383.

### CODE REVIEW

File: truss/truss_config.py [hunk 1]
```diff
@@ -267,6 +267,7 @@ class TrussConfig:
     apply_library_patches: bool = True
     # spec_version is a version string
     spec_version: str = DEFAULT_SPEC_VERSION
+    trainable: bool = False
```
Comment (bolasim) [inline truss/truss_config.py:270]: Doesn't this add an extra key that you'd end up serializing to the config?

Comment (varunshenoy) [reply truss/truss_config.py:270]: I don't think so. The serialization takes place in the `to_dict` method of `TrussConfig` and any keys left out aren't outputted into the .yml.

File: truss/build.py [hunk 2]
```diff
@@ -233,10 +233,12 @@ def init(
     """
     config = TrussConfig(
         model_type="custom",
```
Comment (bolasim) [inline truss/build.py:235]: can we just drop this too and just construct with all the defaults with the exception of the python version?

File: truss/build.py [hunk 3]
```diff
@@ -343,6 +345,8 @@ def _populate_default_training_code(
     # now we don't support this.
     truss_template = "custom"
     template_path = TEMPLATES_DIR / truss_template
+
+    # check if train is valid
```
Comment (bolasim) [inline truss/build.py:349]: not sure I understand this comment..

File: truss/truss_config.py [hunk 4]
```diff
@@ -59,6 +59,8 @@ def to_str(self) -> Optional[str]:
             return f"{self.accelerator.value}:{self.count}"
         return self.accelerator.value
 
+    to_dict = to_str
```
Comment (bolasim) [inline truss/truss_config.py:62]: do we still need these now that you handled those cases better below?

File: truss/truss_config.py [hunk 5]
```diff
@@ -372,3 +349,52 @@ def clone(self):
     def validate(self):
         for secret_name in self.secrets:
             validate_secret_name(secret_name)
+
+
+dataclass_to_req_keys_map = {
```
Comment (bolasim) [inline truss/truss_config.py:354]: Should probably make a constant (ALL_CAPS) and move to the top

File: truss/truss_config.py [hunk 6]
```diff
@@ -372,3 +349,52 @@ def clone(self):
...
+        "python_version",
+        "requirements",
+        "resources",
+        "secrets",
+        "system_packages",
+    },
+    BaseImage: {"image", "python_executable_path"},
+}
+
+types_to_serialize = (Resources, Train, BaseImage)
```
Comment (bolasim) [inline truss/truss_config.py:371]: hows this different than the keys of the object right above? Might be good to write some comments on how you intend to use these

**Review by bolasim — COMMENTED** [review]
LGTM. Just a few minor comments before approving (and to give other people a chance to review).

Also, if you bump the version in pyproject.toml to 0.4.10rc3, not only do you get to merge your first PR, but you also get to release your first python package! you can see all the older released versions here: https://pypi.org/project/truss/#history

File: truss/truss_config.py (hunk 6 above)
Comment (varunshenoy) [reply truss/truss_config.py]: we can just use the keys from the object above

File: truss/truss_config.py (hunk 5 above)
Comment (varunshenoy) [reply truss/truss_config.py]: Can't move it to the top since all the classnames are not defined yet.

File: truss/build.py (hunk 3 above)
Comment (varunshenoy) [reply truss/build.py]: should've been deleted, my bad!

File: truss/tests/test_config.py [hunk 7]
```diff
@@ -109,3 +113,83 @@ def test_parse_base_image(input_dict, expect_base_image, output_dict):
...
+  cpu: 500m
+  memory: 512Mi
+  use_gpu: false
+secrets: {}
+system_packages: []
+"""
+
+    assert config_yaml.strip() == yaml.dump(config.to_dict()).strip()
+
+    config = TrussConfig(
```
Comment (bolasim) [inline truss/tests/test_config.py:141]: nit (optional): these probably could be separate tests or parameterized.

File: truss/truss_config.py [hunk 8]
```diff
@@ -75,6 +75,8 @@ def from_str(acc_spec: Optional[str]):
             raise ValidationError(f"Accelerator {acc_spec} not supported") from exc
         return AcceleratorSpec(accelerator=acc, count=count)
 
+    from_dict = from_str
```
Comment (bolasim) [inline truss/truss_config.py:78]: do westill need this?

**Review by bolasim — APPROVED** [review]
Excellent work!

File: truss/truss_config.py [hunk 9]
```diff
@@ -372,3 +347,50 @@ def clone(self):
...
+        "requirements",
+        "resources",
+        "secrets",
+        "system_packages",
+    },
+    BaseImage: {"image", "python_executable_path"},
+}
+
+
+def obj_to_dict(obj):
```
Comment (joostinyi) [inline truss/truss_config.py:370]: please add a doc string or inline comments so we can more easily parse this when we looks to extend/change the config class

**Review by joostinyi — APPROVED** [review]
nice work!

**Comment by joostinyi** [comment]
can you also verify that new configs default to having `spec_version: 2.0`?

**Comment by bolasim** [comment]
> can you also verify that new configs default to having `spec_version: 2.0`?

We should drop that key if it's default as well. Users don't care.

**Pull request merged**